File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/LevelingIndicator.vb
Imports Sakuno.ING.Game.Models
Imports Windows.UI

Namespace Controls
    <TemplatePart(Name:=LevelingIndicator.Name_PART_Range, Type:=GetType(RangeBase))>
    <TemplatePart(Name:=LevelingIndicator.Name_PART_Level, Type:=GetType(TextBlock))>
    Public Class LevelingIndicator
        Inherits Control
        Friend Const Name_PART_Range = "PART_Range"
        Friend Const Name_PART_Level = "PART_Level"

        Public Sub New()
            DefaultStyleKey = GetType(LevelingIndicator)
        End Sub

        Private _level As Leveling
        Public Property Level As Leveling
            Get
                Return _level
            End Get
            Set(value As Leveling)
                _level = value
                Update()
            End Set
        End Property

        Public Shared ReadOnly TextForegroundProperty As DependencyProperty =
            DependencyProperty.Register(NameOf(TextForeground), GetType(Brush), GetType(LevelingIndicator),
                New PropertyMetadata(New SolidColorBrush(Colors.Transparent)))
        Public Property TextForeground As Brush
            Get
                Return DirectCast(GetValue(TextForegroundProperty), Brush)
            End Get
            Set(value As Brush)
                SetValue(TextForegroundProperty, value)
            End Set
        End Property

        Private PART_Range As RangeBase
        Private PART_Level As TextBlock
        Protected Overrides Sub OnApplyTemplate()
            MyBase.OnApplyTemplate()
            PART_Range = TryCast(GetTemplateChild(Name_PART_Range), RangeBase)
            PART_Level = TryCast(GetTemplateChild(Name_PART_Level), TextBlock)
            Update()
        End Sub

        Private Sub Update()
            Dim l = Level

            If PART_Range IsNot Nothing Then
                If l.CurrentLevelExperience >= l.NextLevelExperience Then
                    PART_Range.Minimum = 0
                    PART_Range.Maximum = 1
                    PART_Range.Value = 1
                Else
                    PART_Range.Minimum = l.CurrentLevelExperience
                    PART_Range.Maximum = l.NextLevelExperience
                    PART_Range.Value = l.Experience
                End If
            End If

            If PART_Level IsNot Nothing Then
                PART_Level.Text = l.Level.ToString()
            End If
        End Sub
    End Class
End Namespace

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/LevelingIndicator.cs
using Sakuno.ING.Game.Models;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Controls.Primitives;
using Windows.UI.Xaml.Media;

namespace Sakuno.ING.Views.UWP.Controls
{
    [TemplatePart(Name = nameof(PART_Range), Type = typeof(RangeBase))]
    [TemplatePart(Name = nameof(PART_Level), Type = typeof(TextBlock))]
    public sealed class LevelingIndicator : Control
    {
        public LevelingIndicator()
        {
            DefaultStyleKey = typeof(LevelingIndicator);
        }

        private Leveling _level;
        public Leveling Level
        {
            get => _level;
            set
            {
                _level = value;
                Update();
            }
        }

        public static DependencyProperty TextForegroundProperty =
            DependencyProperty.Register(nameof(TextForeground), typeof(Brush), typeof(LevelingIndicator),
                new PropertyMetadata(null));
        public Brush TextForeground
        {
            get => (Brush)GetValue(TextForegroundProperty);
            set => SetValue(TextForegroundProperty, value);
        }

        private RangeBase PART_Range;
        private TextBlock PART_Level;
        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            PART_Range = GetTemplateChild(nameof(PART_Range)) as RangeBase;
            PART_Level = GetTemplateChild(nameof(PART_Level)) as TextBlock;
            Update();
        }

        private void Update()
        {
            if (PART_Range != null)
                if (_level.CurrentLevelExperience >= _level.NextLevelExperience)
                {
                    PART_Range.Minimum = 0;
                    PART_Range.Maximum = 1;
                    PART_Range.Value = 1;
                }
                else
                {
                    PART_Range.Minimum = _level.CurrentLevelExperience;
                    PART_Range.Maximum = _level.NextLevelExperience;
                    PART_Range.Value = _level.Experience;
                }

            if (PART_Level != null)
                PART_Level.Text = _level.Level.ToString();
        }
    }
}
